### PR TITLE
Prevent Subscription to attributes with no access privilege and absence

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
Reads and Subscribes from nodes with no access still get serviced to completion (which for subs mean that a subscription is established). This ends up stealing precious resources from nodes that do have access.

Fixes: issue #18485 

#### Change overview
Prevents reads/subs from moving past the processing stage to actual allocation of resources. A StatusResponse is returned immediately.

Caveat: This doesn't address lack of access for event paths. That will come separately. currently we don't have ways to expand wildcard event path and check if path exists, suggest to figure out this in post1.0. At this moment, subscription prevention with no access privilege and absence is the best effort we can go 

#### Testing
add unit test to validate empty attribute paths and event paths in subscribe request
add unit test to validate non-exist concrete attribute path and event paths in subscribe request
add unit test to validate acl-denied attribute path and event path in subscribe request

### CODE REVIEW

File: src/app/InteractionModelEngine.cpp [hunk 1]
```diff
@@ -305,6 +308,58 @@ Status InteractionModelEngine::OnInvokeCommandRequest(Messaging::ExchangeContext
...
+        AttributePathParams params;
+        ReturnErrorOnFailure(path.ParsePath(params));
+
+        //
+        // We create an iterator to point to a single item object list that tracks the path we just parsed.
+        // This avoids the 'parse all paths' approach that is employed in ReadHandler since we want to
+        // avoid allocating out of the path store during this minimal initial processing stage.
+        //
+        ObjectList<AttributePathParams> paramsList;
+        paramsList.mValue = params;
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:338]: Can we not parse the path directly into paramsList.mValue so we use up less stack space and don't have to copy the path?

File: src/app/InteractionModelEngine.cpp [hunk 2]
```diff
@@ -305,6 +308,58 @@ Status InteractionModelEngine::OnInvokeCommandRequest(Messaging::ExchangeContext
...
+        AttributePathExpandIterator pathIterator(&paramsList);
+        ConcreteAttributePath readPath;
+
+        for (; pathIterator.Get(readPath); pathIterator.Next())
+        {
+            Access::RequestPath requestPath{ .cluster = readPath.mClusterId, .endpoint = readPath.mEndpointId };
+            err = Access::GetAccessControl().Check(aSubjectDescriptor, requestPath, RequiredPrivilege::ForReadAttribute(readPath));
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:345]: This does not match the spec.  The spec is doing a bunch of other checks too.

File: src/app/InteractionModelEngine.cpp [hunk 3]
```diff
@@ -305,6 +308,58 @@ Status InteractionModelEngine::OnInvokeCommandRequest(Messaging::ExchangeContext
     return Status::Success;
 }
 
+CHIP_ERROR InteractionModelEngine::ParseAttributePaths(Access::SubjectDescriptor & aSubjectDescriptor,
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:311]: Does the SubjectDescriptor need to be a non-const ref?

File: src/app/InteractionModelEngine.cpp [hunk 4]
```diff
@@ -363,17 +415,37 @@ Protocols::InteractionModel::Status InteractionModelEngine::OnReadInitialRequest
...
-                err = CHIP_NO_ERROR;
+                return Status::InvalidAction;
             }
-            if (err != CHIP_NO_ERROR)
+
+            if (requestedAttributePathCount == 0 && requestedEventPathCount == 0)
             {
+                ChipLogError(InteractionModel,
+                             "Subscription from [%u:" ChipLogFormatX64
+                             "] has zero path for attribute and event. Rejecting request.",
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:429]:
```suggestion
                             "] has no attribute or event paths. Rejecting request.",
```

File: src/app/InteractionModelEngine.cpp [hunk 5]
```diff
@@ -363,17 +415,37 @@ Protocols::InteractionModel::Status InteractionModelEngine::OnReadInitialRequest
...
+            // access so for now, assume that the presence of any path in the event list means they
+            // might have some access to those events.
+            //
+            if (!hasValidAttributePath && requestedEventPathCount == 0)
+            {
+                ChipLogError(InteractionModel,
+                             "Subscription from [%u:" ChipLogFormatX64 "] has no access at all. Rejecting request.",
+                             apExchangeContext->GetSessionHandle()->GetFabricIndex(),
+                             ChipLogValueX64(apExchangeContext->GetSessionHandle()->AsSecureSession()->GetPeerNodeId()));
+                return Status::UnsupportedAccess;
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:446]: This is not what the spec says to do.

File: src/app/InteractionModelEngine.cpp [hunk 6]
```diff
@@ -408,6 +476,72 @@ Protocols::InteractionModel::Status InteractionModelEngine::OnReadInitialRequest
...
+        }
+        else if (err != CHIP_ERROR_END_OF_TLV)
+        {
+            return Status::InvalidAction;
+        }
+
+        if (requestedAttributePathCount == 0 && requestedEventPathCount == 0)
+        {
+            ChipLogError(InteractionModel,
+                         "Subscription from [%u:" ChipLogFormatX64 "] has zero path for attribute and event. Rejecting request.",
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:526]: This is not a subscription.  And note the wording suggestion for the subscription case.

File: src/app/InteractionModelEngine.cpp [hunk 7]
```diff
@@ -408,6 +476,72 @@ Protocols::InteractionModel::Status InteractionModelEngine::OnReadInitialRequest
...
+        // TODO: We don't have an easy way to do a similar 'path expansion' for events to deduce
+        // access so for now, assume that the presence of any path in the event list means they
+        // might have some access to those events.
+        //
+        if (!hasValidAttributePath && requestedEventPathCount == 0)
+        {
+            ChipLogError(InteractionModel, "Read from [%u:" ChipLogFormatX64 "] has no access at all. Rejecting request.",
+                         apExchangeContext->GetSessionHandle()->GetFabricIndex(),
+                         ChipLogValueX64(apExchangeContext->GetSessionHandle()->AsSecureSession()->GetPeerNodeId()));
+            return Status::UnsupportedAccess;
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:542]: This does not match the spec.

File: src/app/InteractionModelEngine.h [hunk 8]
```diff
@@ -373,6 +373,22 @@ class InteractionModelEngine : public Messaging::UnsolicitedMessageHandler,
...
+     *
+     * If there is AT LEAST one concrete path that has sufficient privilege, aHasValidAttributePath will be set to true. Otherwise,
+     * it will be set to false.
+     *
+     * aRequestedAttributePathCount will be updated to reflect the number of attribute paths in the request.
+     *
+     * @retval CHIP_NO_ERROR if no error is encountered.
+     *
+     */
+    CHIP_ERROR ParseAttributePaths(Access::SubjectDescriptor & aSubjectDescriptor,
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.h:388]: Does this use any member state?  If not, can we make this static?

File: src/app/MessageDef/EventPathIB.h [hunk 9]
```diff
@@ -127,6 +127,8 @@ class Parser : public ListParser
      *          #CHIP_ERROR_IM_MALFORMED_EVENT_PATH_IB if the path from the reader is not a valid concrere event path.
      */
     CHIP_ERROR GetEventPath(ConcreteEventPath * const apPath) const;
+
+    CHIP_ERROR ParsePath(EventPathParams & aEvent) const;
```
Comment (bzbarsky-apple) [inline src/app/MessageDef/EventPathIB.h:131]: Document, please.

File: src/app/tests/TestAclAttribute.cpp [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements unit tests for CHIP Interaction Model Read Interaction
```
Comment (bzbarsky-apple) [inline src/app/tests/TestAclAttribute.cpp:21]: Seems pretty generic for the filename....

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/InteractionModelEngine.cpp (hunk 3 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:311]: it should be const ref

File: src/app/InteractionModelEngine.cpp (hunk 4 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:429]: Done, thanks

File: src/app/InteractionModelEngine.cpp (hunk 6 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:526]: corrected, thanks

File: src/app/InteractionModelEngine.cpp (hunk 7 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:542]: what should we return?

File: src/app/InteractionModelEngine.h (hunk 8 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.h:388]: yes, should be static.

File: src/app/MessageDef/EventPathIB.h (hunk 9 above)
Comment (yunhanw-google) [reply src/app/MessageDef/EventPathIB.h:131]: Done, thanks

File: src/app/InteractionModelEngine.h (hunk 8 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.h:388]: Done, thanks

File: src/app/tests/TestAclAttribute.cpp (hunk 10 above)
Comment (yunhanw-google) [reply src/app/tests/TestAclAttribute.cpp:21]: fixed, thanks

File: src/app/InteractionModelEngine.cpp (hunk 1 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:338]: yes, updated, thanks

File: src/app/InteractionModelEngine.cpp (hunk 7 above)
Comment (bzbarsky-apple) [reply src/app/InteractionModelEngine.cpp:557]: Why are we returning an error at all?  It's not invalid to have empty things for a read, per spec, as far as I can tell.

Comment (bzbarsky-apple) [reply src/app/InteractionModelEngine.cpp:557]: In fact, we should not be doing the pre-check for read at all, again as far as I can tell: that's not what the spec says to do.

Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:557]: ok, let me remove pre-check for read.

File: src/app/InteractionModelEngine.cpp (hunk 5 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:478]: removed, thanks

File: src/app/InteractionModelEngine.cpp (hunk 2 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp]: updated, for concrete path, we also need to run existence check in ember storage and acl check as well.

File: src/app/InteractionModelEngine.cpp [hunk 11]
```diff
@@ -323,6 +326,77 @@ Status InteractionModelEngine::OnInvokeCommandRequest(Messaging::ExchangeContext
...
+        {
+            AttributePathExpandIterator pathIterator(&paramsList);
+            ConcreteAttributePath readPath;
+
+            for (; pathIterator.Get(readPath); pathIterator.Next())
+            {
+                Access::RequestPath requestPath{ .cluster = readPath.mClusterId, .endpoint = readPath.mEndpointId };
+                err = Access::GetAccessControl().Check(aSubjectDescriptor, requestPath,
+                                                       RequiredPrivilege::ForReadAttribute(readPath));
+                if (err == CHIP_NO_ERROR)
```
Comment (andy31415) [inline src/app/InteractionModelEngine.cpp:365]: Please add a comment that ignoring err here on failure is intentional: we seem to just want to ensure that some valid path exists and we silently ignore the check error if one exists.

File: src/app/InteractionModelEngine.cpp [hunk 12]
```diff
@@ -323,6 +326,77 @@ Status InteractionModelEngine::OnInvokeCommandRequest(Messaging::ExchangeContext
     return Status::Success;
 }
 
+CHIP_ERROR InteractionModelEngine::ParseAttributePaths(const Access::SubjectDescriptor & aSubjectDescriptor,
```
Comment (andy31415) [inline src/app/InteractionModelEngine.cpp:329]: Naming wise, this says `parse`... but its intent seems to be to set aHasValidAttributePath and aRequestedAttributePathCount. Should we name this `ValidateAttributePaths`?

For output arguments, instead of a bool + sizet, how about a `struct` containing them, so that this is extensible in the future?

File: src/app/InteractionModelEngine.cpp [hunk 13]
```diff
@@ -323,6 +326,77 @@ Status InteractionModelEngine::OnInvokeCommandRequest(Messaging::ExchangeContext
...
+            {
+                Access::RequestPath requestPath{ .cluster  = paramsList.mValue.mClusterId,
+                                                 .endpoint = paramsList.mValue.mEndpointId };
+                ConcreteAttributePath concretePath(paramsList.mValue.mEndpointId, paramsList.mValue.mClusterId,
+                                                   paramsList.mValue.mAttributeId);
+                err = Access::GetAccessControl().Check(aSubjectDescriptor, requestPath,
+                                                       RequiredPrivilege::ForReadAttribute(concretePath));
+                if (err == CHIP_NO_ERROR)
+                {
+                    aHasValidAttributePath = true;
```
Comment (andy31415) [inline src/app/InteractionModelEngine.cpp:384]: should we break here? the other path breaks but this does not.

File: src/app/InteractionModelEngine.cpp [hunk 14]
```diff
@@ -323,6 +326,77 @@ Status InteractionModelEngine::OnInvokeCommandRequest(Messaging::ExchangeContext
...
+
+            for (; pathIterator.Get(readPath); pathIterator.Next())
+            {
+                Access::RequestPath requestPath{ .cluster = readPath.mClusterId, .endpoint = readPath.mEndpointId };
+                err = Access::GetAccessControl().Check(aSubjectDescriptor, requestPath,
+                                                       RequiredPrivilege::ForReadAttribute(readPath));
+                if (err == CHIP_NO_ERROR)
+                {
+                    aHasValidAttributePath = true;
+                    break;
```
Comment (andy31415) [inline src/app/InteractionModelEngine.cpp:368]: how does breaking and aRequestedAttributePathCount interaction work together?

File: src/app/InteractionModelEngine.h [hunk 15]
```diff
@@ -378,6 +378,22 @@ class InteractionModelEngine : public Messaging::UnsolicitedMessageHandler,
...
+    /**
+     * This parses the attribute path list to ensure it is well formed. If so, for each path in the list, it will expand to a list
+     * of concrete paths and walk each path to check if it has privileges to read that attribute.
+     *
+     * If there is AT LEAST one concrete path that has sufficient privilege, aHasValidAttributePath will be set to true. Otherwise,
+     * it will be set to false.
+     *
+     * aRequestedAttributePathCount will be updated to reflect the number of attribute paths in the request.
+     *
+     * @retval CHIP_NO_ERROR if no error is encountered.
```
Comment (andy31415) [inline src/app/InteractionModelEngine.h:390]: This `@retval` is obvious. I would remove this from the documentation. anything returning CHIP_ERROR is expected to return CHIP_NO_ERROR on success and an error code on failure.

**Review by andy31415 — COMMENTED** [review]
@yunhanw-google why was a large chunk of code regarding `EnsureResourceForRead` deleted and overall larger deltas in the PR?

The summary description says adding checks, which I can see as `ParseAttributePaths`, but why the other deltas?

File: src/app/MessageDef/AttributePathIB.cpp [hunk 16]
```diff
@@ -227,6 +227,62 @@ CHIP_ERROR AttributePathIB::Parser::GetListIndex(ConcreteDataAttributePath & aAt
...
+    if (err == CHIP_NO_ERROR)
+    {
+        VerifyOrReturnError(!aAttribute.HasWildcardAttributeId() && !aAttribute.HasWildcardListIndex(),
+                            CHIP_IM_GLOBAL_STATUS(InvalidAction));
+    }
+    else if (err == CHIP_END_OF_TLV)
+    {
+        err = CHIP_NO_ERROR;
+    }
+    return err;
```
Comment (andy31415) [inline src/app/MessageDef/AttributePathIB.cpp:283]: The documentation of the method says that `InvalidAction` is always returned on failure.
This seems incorrect:
  - this return is the value from GetListIndex (may be any error)
  - Line 265 returns error on failure same with line 254 - so any error code could be possible. Just InvalidAction is the more common one.

File: src/app/MessageDef/AttributePathIB.h [hunk 17]
```diff
@@ -153,6 +153,14 @@ class Parser : public ListParser
     CHIP_ERROR GetListIndex(ConcreteDataAttributePath & aAttributePath) const;
 
     // TODO(#14934) Add a function to get ConcreteDataAttributePath from AttributePathIB::Parser directly.
+
+    /**
+     * @brief Parse the attribute path into an AttributePathParams object. As part of parsing,
+     * validity checks for each path item will be done as well.
+     *
+     * If any errors are encountered, an IM error of 'InvalidAction' will be returned.
```
Comment (andy31415) [inline src/app/MessageDef/AttributePathIB.h:161]: This method can return more than just InvalidAction. Maybe restate as `validation errors result in InvalidAction. Other parser errors may result in different error codes`.

File: src/app/tests/TestAclAttribute.cpp [hunk 18]
```diff
@@ -0,0 +1,294 @@
...
+
+chip::ClusterId kTestClusterId        = 1;
+chip::ClusterId kTestDeniedClusterId1 = 2;
+chip::ClusterId kTestDeniedClusterId2 = 3;
+chip::EndpointId kTestEndpointId      = 4;
+
+class TestAccessControlDelegate : public AccessControl::Delegate
+{
+public:
+    // Check
```
Comment (andy31415) [inline src/app/tests/TestAclAttribute.cpp:62]: Delte this comment? it seems to just restate the name of the method below.

File: src/app/InteractionModelEngine.cpp (hunk 7 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp]: @andy31415 we don't need precheck for read per spec

**Comment by yunhanw-google** [comment]
> @yunhanw-google why was a large chunk of code regarding `EnsureResourceForRead` deleted and overall larger deltas in the PR?
> 
> 
> 
> The summary description says adding checks, which I can see as `ParseAttributePaths`, but why the other deltas?

Please see this thread,
https://github.com/project-chip/connectedhomeip/pull/22349#discussion_r961956600, per @bzbarsky-apple the original check for read is not needed and not aligned with spec

File: src/app/InteractionModelEngine.cpp (hunk 11 above)
Comment (bzbarsky-apple) [reply src/app/InteractionModelEngine.cpp:365]: The point is, the definitio nof "valid path" is "path exists and ACL allows access".  The "path exists" part is handled by AttributePathExpandIterator.  So we just need to check the ACL bits.

File: src/app/InteractionModelEngine.cpp (hunk 14 above)
Comment (bzbarsky-apple) [reply src/app/InteractionModelEngine.cpp:368]: aRequestedAttributePathCount is counting paths, not concrete paths they expand into.

File: src/app/InteractionModelEngine.cpp (hunk 13 above)
Comment (bzbarsky-apple) [reply src/app/InteractionModelEngine.cpp:384]: The break above is from the concrete path iterator loop, not the loop over request paths....

File: src/app/InteractionModelEngine.cpp [hunk 19]
```diff
@@ -381,17 +452,36 @@ Protocols::InteractionModel::Status InteractionModelEngine::OnReadInitialRequest
...
+            // access so for now, assume that the presence of any path in the event list means they
+            // might have some access to those events.
+            //
+            if (!hasValidAttributePath && requestedEventPathCount == 0)
+            {
+                ChipLogError(InteractionModel,
+                             "Subscription from [%u:" ChipLogFormatX64 "] has no access at all. Rejecting request.",
+                             apExchangeContext->GetSessionHandle()->GetFabricIndex(),
+                             ChipLogValueX64(apExchangeContext->GetSessionHandle()->AsSecureSession()->GetPeerNodeId()));
+                return Status::UnsupportedAccess;
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:482]: Again, this is not the right error status for this situation, per spec.

File: src/app/InteractionModelEngine.cpp [hunk 20]
```diff
@@ -427,54 +513,6 @@ Protocols::InteractionModel::Status InteractionModelEngine::OnReadInitialRequest
...
-                ReturnErrorCodeIf(TLV::Utilities::Count(pathReader, requestedEventPathCount, false) != CHIP_NO_ERROR,
-                                  Status::InvalidAction);
-            }
-            else if (err != CHIP_ERROR_END_OF_TLV)
-            {
-                return Status::InvalidAction;
-            }
-
-            // The following cast is safe, since we can only hold a few tens of paths in one request.
-            Status checkResult = EnsureResourceForRead(apExchangeContext->GetSessionHandle()->GetFabricIndex(),
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:469]: Wait, what?  Why is `EnsureResourceForRead` being removed?  That's not right!

File: src/app/InteractionModelEngine.h [hunk 21]
```diff
@@ -378,6 +378,22 @@ class InteractionModelEngine : public Messaging::UnsolicitedMessageHandler,
                                  System::PacketBufferHandle && aPayload) override;
     void OnResponseTimeout(Messaging::ExchangeContext * ec) override;
 
+    /**
+     * This parses the attribute path list to ensure it is well formed. If so, for each path in the list, it will expand to a list
+     * of concrete paths and walk each path to check if it has privileges to read that attribute.
+     *
+     * If there is AT LEAST one concrete path that has sufficient privilege, aHasValidAttributePath will be set to true. Otherwise,
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.h:385]:
```suggestion
     * If there is AT LEAST one "existent path" (as the spec calls it) that has sufficient privilege, aHasValidAttributePath will be set to true. Otherwise,
```

File: src/app/util/ember-compatibility-functions.cpp [hunk 22]
```diff
@@ -516,6 +516,27 @@ Protocols::InteractionModel::Status UnsupportedAttributeStatus(const ConcreteAtt
...
 } // anonymous namespace
 
+bool CheckConcreteAttributePathExistence(const AttributePathParams & aPath)
+{
+    const EmberAfAttributeMetadata * attributeMetadata = nullptr;
+    const EmberAfCluster * attributeCluster            = nullptr;
+
+    switch (aPath.mAttributeId)
+    {
+    case Clusters::Globals::Attributes::AttributeList::Id:
```
Comment (bzbarsky-apple) [inline src/app/util/ember-compatibility-functions.cpp:526]: Please use `GlobalAttributesNotInMetadata` from `src/app/GlobalAttributes.h` instead of hardcoding a second copy of that list.

We should fix `ReadSingleClusterData` to do that too, in a follow-up.

File: src/app/util/ember-compatibility-functions.cpp [hunk 23]
```diff
@@ -516,6 +516,27 @@ Protocols::InteractionModel::Status UnsupportedAttributeStatus(const ConcreteAtt
 
 } // anonymous namespace
 
+bool CheckConcreteAttributePathExistence(const AttributePathParams & aPath)
```
Comment (bzbarsky-apple) [inline src/app/util/ember-compatibility-functions.cpp:519]:
```suggestion
bool ConcreteAttributePathExists(const AttributePathParams & aPath)
```

File: src/app/util/ember-compatibility-functions.cpp [hunk 24]
```diff
@@ -516,6 +516,27 @@ Protocols::InteractionModel::Status UnsupportedAttributeStatus(const ConcreteAtt
...
+    case Clusters::Globals::Attributes::AcceptedCommandList::Id:
+        FALLTHROUGH;
+    case Clusters::Globals::Attributes::GeneratedCommandList::Id:
+        attributeCluster = emberAfFindCluster(aPath.mEndpointId, aPath.mClusterId, CLUSTER_MASK_SERVER);
+        break;
+    default:
+        attributeMetadata = emberAfLocateAttributeMetadata(aPath.mEndpointId, aPath.mClusterId, aPath.mAttributeId);
+    }
+
+    return (attributeCluster == nullptr && attributeMetadata == nullptr) ? false : true;
```
Comment (bzbarsky-apple) [inline src/app/util/ember-compatibility-functions.cpp:537]: How about:
```suggestion
    for (auto & attr : GlobalAttributesNotInMetadata)
    {
        if (attr == aPath.mAttributeId) 
        {
            return (emberAfFindCluster(aPath.mEndpointId, aPath.mClusterId, CLUSTER_MASK_SERVER) != nullptr);
        }
    }
    return (emberAfLocateAttributeMetadata(aPath.mEndpointId, aPath.mClusterId, aPath.mAttributeId) != nullptr);
```

File: src/app/util/ember-compatibility-functions.cpp [hunk 25]
```diff
@@ -542,7 +563,6 @@ CHIP_ERROR ReadSingleClusterData(const SubjectDescriptor & aSubjectDescriptor, b
     default:
         attributeMetadata = emberAfLocateAttributeMetadata(aPath.mEndpointId, aPath.mClusterId, aPath.mAttributeId);
     }
-
```
Comment (bzbarsky-apple) [inline src/app/util/ember-compatibility-functions.cpp:545]: Why this change?

Comment (yunhanw-google) [reply src/app/util/ember-compatibility-functions.cpp:545]: accidental removal, updated. thanks

File: src/app/util/ember-compatibility-functions.cpp (hunk 24 above)
Comment (yunhanw-google) [reply src/app/util/ember-compatibility-functions.cpp:537]: looks better, updated, thanks

File: src/app/util/ember-compatibility-functions.cpp (hunk 23 above)
Comment (yunhanw-google) [reply src/app/util/ember-compatibility-functions.cpp:519]: updated, thanks

File: src/app/util/ember-compatibility-functions.cpp (hunk 22 above)
Comment (yunhanw-google) [reply src/app/util/ember-compatibility-functions.cpp:526]: updated, would follow-up in ReadSingleClusterData
https://github.com/project-chip/connectedhomeip/pull/22349

File: src/app/InteractionModelEngine.h (hunk 21 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.h:385]: updated, thanks

File: src/app/InteractionModelEngine.cpp (hunk 20 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:469]: just bring it up, thanks for pointing out this.

File: src/app/InteractionModelEngine.cpp (hunk 19 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp]: InvalidAction?

File: src/app/InteractionModelEngine.cpp (hunk 12 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:329]: actually we have not yet validate if the paths are good to go, the upper layer uses the returned bool & aHasValidAttributePath, size_t & aRequestedAttributePathCount and others to validate if the request is roughly good or not

File: src/app/InteractionModelEngine.h (hunk 15 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.h:390]: removed. thanks

File: src/app/MessageDef/AttributePathIB.cpp (hunk 16 above)
Comment (yunhanw-google) [reply src/app/MessageDef/AttributePathIB.cpp:283]: yes, invalidaction is most common one.

Comment (yunhanw-google) [reply src/app/MessageDef/AttributePathIB.cpp:283]: just update the code, and return unified invalid action for all error cases. thanks

File: src/app/MessageDef/EventPathIB.h [hunk 26]
```diff
@@ -127,6 +127,14 @@ class Parser : public ListParser
      *          #CHIP_ERROR_IM_MALFORMED_EVENT_PATH_IB if the path from the reader is not a valid concrere event path.
      */
     CHIP_ERROR GetEventPath(ConcreteEventPath * const apPath) const;
+
+    /**
+     * @brief Parse the event path into an EventPathParams object. As part of parsing,
+     * validity checks for each path item will be done as well.
+     *
+     * If any errors are encountered, an IM error of 'InvalidAction' will be returned.
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/MessageDef/EventPathIB.h:135]: updated, thanks

File: src/app/tests/TestAclAttribute.cpp [hunk 27]
```diff
@@ -0,0 +1,294 @@
...
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/tests/TestAclAttribute.cpp:20]: removed, thanks

File: src/app/tests/TestAclAttribute.cpp (hunk 18 above)
Comment (yunhanw-google) [reply src/app/tests/TestAclAttribute.cpp:62]: removed, thanks

File: src/app/tests/TestAclAttribute.cpp [hunk 28]
```diff
@@ -0,0 +1,294 @@
...
+    CHIP_ERROR mError = CHIP_NO_ERROR;
+};
+} // namespace
+
+namespace chip {
+namespace app {
+
+bool CheckConcreteAttributePathExistence(const AttributePathParams & aPath)
+{
+    if (aPath.mClusterId == kTestDeniedClusterId1)
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/tests/TestAclAttribute.cpp:147]: updated, thanks

File: src/app/MessageDef/AttributePathIB.h (hunk 17 above)
Comment (yunhanw-google) [reply src/app/MessageDef/AttributePathIB.h:161]: updated the code, now it all return invalid action.

File: src/app/InteractionModelEngine.cpp [hunk 29]
```diff
@@ -381,14 +454,33 @@ Protocols::InteractionModel::Status InteractionModelEngine::OnReadInitialRequest
             {
                 TLV::TLVReader pathReader;
                 eventpathListParser.GetReader(&pathReader);
-                err = TLV::Utilities::Count(pathReader, requestedEventPathCount, false);
+                TLV::Utilities::Count(pathReader, requestedEventPathCount, false);
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:457]: Why is this no longer checking the returned CHIP_ERROR?  It should be checking it, I would think.

File: src/app/InteractionModelEngine.h [hunk 30]
```diff
@@ -378,6 +378,21 @@ class InteractionModelEngine : public Messaging::UnsolicitedMessageHandler,
                                  System::PacketBufferHandle && aPayload) override;
     void OnResponseTimeout(Messaging::ExchangeContext * ec) override;
 
+    /**
+     * This parses the attribute path list to ensure it is well formed. If so, for each path in the list, it will expand to a list
+     * of concrete paths and walk each path to check if it has privileges to read that attribute.
+     *
+     * If there is AT LEAST one "existent path" (as the spec calls it)" path that has sufficient privilege, aHasValidAttributePath
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.h:385]:
```suggestion
     * If there is AT LEAST one "existent path" (as the spec calls it) that has sufficient privilege, aHasValidAttributePath
```

File: src/app/InteractionModelEngine.h [hunk 31]
```diff
@@ -617,6 +632,13 @@ CHIP_ERROR ReadSingleClusterData(const Access::SubjectDescriptor & aSubjectDescr
                                  const ConcreteReadAttributePath & aPath, AttributeReportIBs::Builder & aAttributeReports,
                                  AttributeValueEncoder::AttributeEncodeState * apEncoderState);
 
+/**
+ *  Check whether concreate attribute path is in ember storage.
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.h:636]:
```suggestion
 *  Check whether concrete attribute path is an "existent attribute path" in spec terms.
```

It might not be in "ember storage" at all; it might be external. This is just checking for _metadata_, and not even that for the global ones.

File: src/darwin/Framework/CHIP/MTRIMDispatch.mm [hunk 32]
```diff
@@ -121,6 +121,11 @@ CHIP_ERROR ReadSingleClusterData(const SubjectDescriptor & aSubjectDescriptor, b
         return aAttributeReports.EncodeAttributeStatus(aPath, StatusIB(status));
     }
 
+    bool ConcreteAttributePathExists(const AttributePathParams & aPath)
+    {
+        return DetermineAttributeStatus(aPath, /* aIsWrite = */ false) == Status::Success;
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRIMDispatch.mm:126]: Just realized this bit is here.  This is not really right, not least because `DetermineAttributeStatus` _always_ returns a failure status.  This should probably be more like:
```suggestion
        return DetermineAttributeStatus(aPath, /* aIsWrite = */ false) == Status::UnsupportedAccess;
```

File: src/app/InteractionModelEngine.h [hunk 33]
```diff
@@ -617,6 +632,13 @@ CHIP_ERROR ReadSingleClusterData(const Access::SubjectDescriptor & aSubjectDescr
...
+ *  @param[in]    aPath                 The concrete path of the data being read.
+ *  @retval  boolean
+ */
+bool ConcreteAttributePathExists(const AttributePathParams & aPath);
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.h:640]: Also, shouldn't this take a `ConcreteAttributePath`?  Or at least document that this should only be called if the `AttributePathParams` represents a concrete path or something?

File: src/app/InteractionModelEngine.h (hunk 31 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.h:636]: updated, thanks

File: src/app/InteractionModelEngine.h (hunk 30 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.h:385]: updated, thanks

File: src/app/InteractionModelEngine.cpp (hunk 29 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:457]: agree, updated, thanks

Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:457]: updated with                 ReturnErrorCodeIf(TLV::Utilities::Count(pathReader, requestedEventPathCount, false) != CHIP_NO_ERROR,   Status::InvalidAction);

File: src/app/InteractionModelEngine.h [hunk 34]
```diff
@@ -378,6 +378,21 @@ class InteractionModelEngine : public Messaging::UnsolicitedMessageHandler,
                                  System::PacketBufferHandle && aPayload) override;
     void OnResponseTimeout(Messaging::ExchangeContext * ec) override;
 
+    /**
+     * This parses the attribute path list to ensure it is well formed. If so, for each path in the list, it will expand to a list
+     * of concrete paths and walk each path to check if it has privileges to read that attribute.
+     *
+     * If there is AT LEAST one "existent path" (as the spec calls it)" that has sufficient privilege, aHasValidAttributePath
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.h:385]:
```suggestion
     * If there is AT LEAST one "existent path" (as the spec calls it) that has sufficient privilege, aHasValidAttributePath
```

**Review by bzbarsky-apple — APPROVED** [review]

**Review by yufengwangca — APPROVED** [review]

**Comment by andy31415** [comment]
1.0 accepted: this is fixing a 1.0 release blocker

**Review by msandstedt — APPROVED** [review]
